utils/fasm/test/test_fasm.cpp

```diff
@@ -4,16 +4,173 @@
 #include "vtr_util.h"
 #include "rr_metadata.h"
 #include "fasm.h"
+#include "fasm_utils.h"
 #include "arch_util.h"
 #include "rr_graph_writer.h"
 #include <sstream>
+#include <fstream>
+#include <regex>
+#include <cmath>
+#include <algorithm>
 
 static constexpr const char kArchFile[] = "test_fasm_arch.xml";
 static constexpr const char kRrGraphFile[] = "test_fasm_rrgraph.xml";
 
+// Prototype here to be able to test it
+namespace fasm {
+std::vector<std::string> find_tags_in_feature (const std::string& a_String);
+}
+
 namespace {
 
+// ============================================================================
+
 using Catch::Matchers::Equals;
+using Catch::Matchers::StartsWith;
+using Catch::Matchers::Contains;
+
+class RegexMatcher : public Catch::MatcherBase<std::string> {
+
+    std::string m_RegexStr;
+    std::regex  m_Regex;
+
+public:
+
+    RegexMatcher (const std::string& a_Regex) : 
+        m_RegexStr(a_Regex), m_Regex(a_Regex) {}
+
+    bool match(const std::string& str) const override {
+        std::smatch match;
+        return std::regex_match(str, match, m_Regex);
+    }
+
+    virtual std::string describe() const override {
+        return "does not match the regex '" + m_RegexStr + "'";
+    }
+};
+
+inline RegexMatcher MatchesRegex (const std::string& a_Regex) {
+    return RegexMatcher(a_Regex);
+}
+
+// ============================================================================
+
+
+TEST_CASE("find_tags", "[fasm]") {
+    const std::string feature ("{prefix}CLB_{loc}_{site}");
+
+    auto tags = fasm::find_tags_in_feature (feature);
+
+    REQUIRE (tags.size() == 3);
+    REQUIRE (std::find(tags.begin(), tags.end(), "prefix") != tags.end());
+    REQUIRE (std::find(tags.begin(), tags.end(), "loc")    != tags.end());
+    REQUIRE (std::find(tags.begin(), tags.end(), "site")   != tags.end());
+}
+
+TEST_CASE("substitute_tags_correct", "[fasm]") {
+    const std::string feature ("{prefix}CLB_{loc}_{site}");
+    const std::map<const std::string, std::string> tags = {
+        {"prefix", "L"},
+        {"loc", "X7Y8"},
+        {"site", "SLICE"}
+    };
+
+    auto result = fasm::substitute_tags(feature, tags);
+
+    REQUIRE(result.compare("LCLB_X7Y8_SLICE") == 0);
+}
+
+TEST_CASE("substitute_tags_undef", "[fasm]") {
+    const std::string feature ("{prefix}CLB_{loc}_{site}");
+    const std::map<const std::string, std::string> tags = {
+        {"loc", "X7Y8"},
+        {"site", "SLICE"}
+    };
+
+    REQUIRE_THROWS(fasm::substitute_tags(feature, tags));
+}
+
+// ============================================================================
+
+/*
+VPR may rearrange LUT inputs and modify its content accordingly. This function
```

Comment:
You shouldn't need to brute force the permutations, as VPR explicitly indicates how the LUTs are rotated.  See https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/master/utils/fasm/src/fasm.cpp#L316

Reply:
I know that the permutation can be extracted from VPR and this is what the genfasm does. But I see no point in duplicating the LUT extraction code (being tested) just for the purpose of the test. As in the end we might not know where a potential bug is (in the code or in the test).

Reply:
We've relied on our logical equivalency check integration tests (strong/strong_func*) to verify that the LUT permutations VTR has stored are correct. These tests use ABC to verify the input netlist to VPR and the post-synthesis netlist (generated by VPR which has the input rotations applied) are equivalent.